Adds CADC download option, plus fixes bugs.

Nat1405 opened the pull request:
Users can now specify -c (eg; runNifty nifsPipeline -c ...)
to download raw data from the Canadian Astronomy Data Centre.
This has been tested to work from the interactive config session
(runNifty nifsPipeline -i) and the fully automatic mode
(runNifty nifsPipeline -c -f <PROGRAM ID>).

Nat1405 committed:
Adds CADC download option, plus fixes bugs.

Users can now specify -c (eg; runNifty nifsPipeline -c ...)
to download raw data from the Canadian Astronomy Data Centre.
This has been tested to work from the interactive config session
(runNifty nifsPipeline -i) and the fully automatic mode
(runNifty nifsPipeline -c -f <PROGRAM ID>).

ijiraq commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1163,3 +1185,47 @@ def MEFarith(MEF, image, op, result):
+    urls = cadc.get_data_urls(result)
+    for url, pid in zip(urls, pids):
+        try:
+            urllib.urlretrieve(url, directory+'/'+pid+'.fits')
```
In stead of urllib can you use 'requests' .  That's a more abstracted interface that provides some extra robustness against changes in the lower library.  Also, this would allow you to look in the `header` of the HTTP `response` to see what the file name should be.  `requests` is part of the astroquery dependencies already.

Here is an example header: 
```Date: Thu, 07 May 2020 16:28:23 GMT
Strict-Transport-Security: max-age=0
Content-MD5: d13af3725015ebd2babd09a8f2eaac3d
ETag: d13af3725015ebd2babd09a8f2eaac3d
Content-Encoding: x-fits
Last-Modified: Mon, 04 Mar 2019 08:09:22 GMT
Content-Disposition: inline; filename=2376828o.fits.fz
X-Uncompressed-Length: 786038400
X-Uncompressed-MD5: 988909a91ab429bbb8a5f51b3d35715e
X-File-CRC: 
X-Uncompressed-CRC: 
Content-Type: application/fits
Content-Length: 330462720
Vary: Origin
Connection: close```
`Content-Disposition` can then be used to name the file.

Nat1405 replied:
Good idea, see 69e03fd2 . Maybe this is what you're thinking.

ijiraq replied:
Yes, that looks better.   I made some comments in the commit.

Nat1405 replied:
Perfect, thank you for all your feedback. Perhaps bbf9efe fixes this.

ijiraq commented on nifty/pipeline/objectoriented/GetConfig.py:
```diff
@@ -72,6 +72,8 @@ def makeConfig(self):
         # Ability to repeat the last data reduction
         self.parser.add_argument('-r', '--repeat', dest = 'repeat', default = False, action = 'store_true', help = 'Repeat the last data reduction, loading saved reduction parameters from runtimeData/config.cfg.')
+        # Specify where downloads come from; either Gemini or CADC.
+        self.parser.add_argument('-c', '--cadc', dest = 'cadc', default = False, action = 'store_true', help = 'Download raw data from Canadian Astronomy Data Centre rather than the Gemini Science Archive.')
```
Hmm.  perhaps --data-source with two choices 'CADC' and 'GSA'  (Gemini Science Archive) would make more sense here.  With GSA being the default.  Eventually you could extend NIFTY with other data sources.

Nat1405 replied:
74a11e9 implements this by changing the "-c" flag to a "-d/--data-source" option with 'GSA' and 'CADC" options. It also adds some better error handling in regards to your next comment.

andamian replied:
`-d` is typically used for debug mode, at least we've been using it for that.

Nat1405 replied:
How about -s/--data-source? Implemented in 9214743 but can be changed very easily in one line.

ijiraq commented on nifty/pipeline/objectoriented/GetConfig.py:
```diff
@@ -121,3 +124,14 @@ def makeConfig(self):
+
+        if self.cadc:
+            try:
+                with open('./' + self.configFile, 'r') as self.config_file:
```
What happens here if the config file doesn't already exist?

Nat1405 replied:
See 74a11e9 for better error handling.

ijiraq commented on nifty/pipeline/steps/nifsSort.py:
```diff
@@ -141,18 +142,22 @@ def start():
+            download_query_cadc(program, os.getcwd()+'/rawData')
         else:
-            download_query_gemini(url, './rawData')
+            url = 'https://archive.gemini.edu/download/'+ str(program) + '/notengineering/NotFail/present/canonical'
```
Is putting this URL building information into the download_query_gemini  more appropriate now?  To mirror the download_query_cadc?   

For the CADC one, we reroute the user to the GSA if we don't have the file, because proprietary.  Passing the proprietary key into download_query_cadc would enable that.  (would require a try-except on the query, and a check to see if permission denied and was the file being asked for from Gemini, or the cookie could just get added to all requests).

Nat1405 replied:
6b976a1 adds the URL building information into the download_query_gemini() function.

Hopefully I understand what you're looking for as actionables from your second point:
- In download_query_cadc(), if the file is not found with a CADC query (even if the user requested that), we'll retry with a proprietary Gemini download?

ijiraq commented on setup.py:
```diff
@@ -22,12 +22,12 @@
-    author_email='mbussero@gemini.edu',
+    version="2.0.0",
+    author='ncomeau',
+    author_email='ncomeau@uvic.ca',
```
Now, we need to decide about where this project should live.  Its true we don't what to bother gemini with mistakes we might be adding, but we also don't want to take credit for code they have written.

Nat1405 replied:
I'll avoid pushing it to PyPI for now. I know some Dockerfiles will eventually depend on this; perhaps we could bundle a built python wheel (.whl) with our Dockerfiles rather than having them pull the package from PyPI? I'll probably eventually have to talk to Marie at Gemini North someday too...

andamian replied:
`Dockerfile` can install from a `github` repo if required.

Nat1405 replied:
Perfect, implemented in https://github.com/ijiraq/gemini_processing/commit/a05258733c2afc26016839a9e290fe0b71b0c1c6 . Now the Dockerfile installs from the "dev" branch of Nifty4Gemini.

ijiraq reviewed (changes requested):
A few changes requested.  Mostly around the idea of how to retrieve the file and be sure the filename is correct.

Nat1405 committed:
Adds support for downloads using requests.

Nat1405 committed:
Splits out getting of file to new method.

Nat1405 committed:
Fixes function naming conventions to match original project.

Nat1405 committed:
Changes '-c' CADC flag to '-d/--data-source' option.

Rather than just choosing between Gemini and CADC downloads,
--data-source adds support for more than just two data sources.
To add support for a new archive at minimum nifsSort.py needs
changed.

Nat1405 committed:
Cleans up url construction code.

Nat1405 committed:
Fixes silly bugs introduced by 74a11e9.

Nat1405 committed:
Adds better error handling to get_file().

Nat1405 committed:
Fixes silly bugs in d5ead22e7.

There were problems with finding the config file.

andamian commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -456,6 +463,21 @@ def interactiveNIFSInput():
+                if isinstance(v, collections.Mapping):
+                    u[k] = update(u.get(k))
+                else:
+                    if u[k] == 'yes':
```
Could use `u[k] = u[k] == 'yes'`

Nat1405 replied:
Great suggestion, very clean. I don't this works here (if I understand what you're suggesting) because u[k] can take on more values than just 'yes' and 'no'. This function is looking to replace all occurrences of 'yes' and 'no' in the dictionary with True and False and leave all other values intact.

andamian commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1163,3 +1185,66 @@ def MEFarith(MEF, image, op, result):
+        # https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/data/pub/GEM/N20140505S0114.fits?RUNID=mf731ukqsipqpdgk
+        filename = (url.split('/')[-1]).split('?')[0]
+    # Write the fits file
+    with open(filename, 'wb') as f:
```
I would probably think writing to a temporary directory or an application directory. You could expect some failures while downloading the file potentially leaving incomplete files behind. Or write them straight into the destination directory with a temporary name (starting with '.' for example). This is where you'd want to compute the md5 of the received data and compare it with the `content-MD5` of the file (assuming the GEMINI also supports it).

Nat1405 replied:
Very good idea. I went with writing to a '.temp-<filename>' file in the working directory for now and doing the md5 check there. I could change it to write the temp file in the download directory as well. Changes are in ff758ba and cf023d9.

andamian replied:
Is there a mechanism to get rid of the failed temporary files? Are they deleted when errors occur? Otherwise they might pile up consuming user's quota. You could delete the old ones on startup or tear down. Might want to do that only for older files if multiple instances of the script are executed at the same time. Alternatively, you could use something like the `tempfile` package to store them in a temporary directory. The cleanup is someone else's job in that case.
'.temp-' is used in multiple places and should be a const. To avoid misspelling bugs.

ijiraq replied:
I recommend using `tempfile.TemporaryFile(mode='w+b', buffering=None, encoding=None, newline=None, suffix=None, prefix=None, dir=None, *, errors=None)`

Set the dir=$CACHEDIR and prefix=filename 

On success mv the file to the correct filename ... if fail raise a exception.  file cleanup is handled by python.

andamian commented on nifty/pipeline/objectoriented/GetConfig.py:
```diff
@@ -60,6 +60,22 @@ def __init__(self, args, script):
+        Checks that a config file exists and if not, sets Nifty to use default configuration.
+        """
+        if not os.path.exists(configFile):
+            shutil.copy(self.RECIPES_PATH+'defaultConfig.cfg', configFile)
```
It's good practice to use `os.path.join()` to create file paths. Applicable to other places.

Nat1405 replied:
Very good catch, thank you. Maybe not doing this inadvertently breaks support for Windows users... I've fixed the calls in GetConfig.py in ad9c64d and [created an issue](https://github.com/Nat1405/Nifty4Gemini/issues/8) to go through the rest of the code base to fix those calls.

Nat1405 committed:
Fixes path seperators (wasn't using os.path.join) in GetConfig.py.

Nat1405 committed:
Sends downloaded CADC files to a temp location first.

Nat1405 committed:
Adds CADC downloads MD5 verification.

Nat1405 committed:
Changes -d/--data-source flag to -s/--data-source flag

Nat1405 committed:
Moves temp files to download directory.

Nat1405 commented:
Fantastic @andamian , thank you for the review. I've pushed draft issues to all the things you've raised.

andamian commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1232,6 +1232,19 @@ def getFile(url):
+    try:
+        server_checksum = r.headers['Content-MD5']
+        with open(filename, 'rb') as f:
+            download_checksum = hashlib.md5(f.read()).hexdigest()
```
GEMINI files are not that big but in general this could be an expensive (e.g. time consuming) operation. Best would be to feed the bytes to `hashlib` at the same time you are writing them into the file (line 1233). That's if the `Content-MD5` present.

Nat1405 replied:
Awesome, good idea. Added in b88d045.

Nat1405 committed:
Changes md5 verification to happen at the same time as downloads.

Nat1405 committed:
Makes CADC downloads use temp files (via tempfile.TemporaryFile).

Nat1405 commented:
Thanks again for another stage of review. I've pushed two new commits, 
b88d045 and 977a915 that address issues raised.

Nat1405 committed:
Adds backwards compatibility for old config files for dataSource option.
